src/v/iceberg/metadata_query.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+#include "base/seastarx.h"
+#include "iceberg/manifest_io.h"
+#include "iceberg/table_metadata.h"
+
+#include <seastar/core/future.hh>
+
+namespace iceberg {
+
+/**
+ * Enum class allowing changing an expected type returned from query
+ */
+enum class result_type { snapshot, manifest_file, manifest };
+
+namespace internal {
+template<result_type T>
+struct results_container {};
+
+template<>
+struct results_container<result_type::snapshot> {
+    using type = chunked_vector<snapshot>;
+};
+
+template<>
+struct results_container<result_type::manifest_file> {
+    using type = chunked_vector<manifest_file>;
+};
+
+template<>
+struct results_container<result_type::manifest> {
+    using type = chunked_vector<manifest>;
+};
+
+} // namespace internal
+
+template<typename ElemT>
+using matcher = ss::noncopyable_function<bool(const ElemT&)>;
```

Comment:
Consider using absl::FunctionRef- it won't allocate and is the `std::string_view` version of `std::function`.

https://github.com/abseil/abseil-cpp/blob/master/absl/functional/function_ref.h